### PR TITLE
Box3 now supports computing minimal bounds for setFromObject

### PR DESCRIPTION
**Pull request by arikwex**
This is a draft for implementing the "minimal bounding box" idea proposed here: https://github.com/mrdoob/three.js/pull/17940#issuecomment-612133803.

I'll need to update the docs and such still, but I'd like to know if everyone's cool with the direction before going further.

### CODE REVIEW

File: src/math/Box3.js [hunk 1]
```diff
@@ -248,24 +248,39 @@ Object.assign( Box3.prototype, {
...
-			if ( geometry.boundingBox === null ) {
+			if ( minimal && geometry.attributes != undefined && geometry.attributes.position !== undefined ) {
 
-				geometry.computeBoundingBox();
+				const position = geometry.attributes.position;
+				for ( let i = 0, l = position.count; i < l; i ++ ) {
 
-			}
+					_vector.set( position.getX( i ), position.getY( i ), position.getZ( i ) );
+					_vector.applyMatrix4( object.matrixWorld );
```
Comment (WestLangley) [inline src/math/Box3.js:257]:
```js
_vector.fromBufferAttribute( position, i ).applyMatrix4( object.matrixWorld );
```

Comment (arikwex) [reply src/math/Box3.js]: !! 🤯 !!
done.

**Comment by arikwex** [comment]
P.S. This does not support morphAttributesPosition in the same manner that BufferGeometry's bounding box computation does. It could be added if necessary.

**Comment by arikwex** [comment]
@WestLangley Am I supposed to do something to nudge this PR? I'm not sure exactly what the expected PR lifecycle is for this ThreeJS.

**Comment by WestLangley** [comment]
@arikwex I added a milestone -- that should help prevent this PR from getting lost -- but I have no control over what gets merged.

**Comment by arikwex** [comment]
Dear @mrdoob, should I refer to you as
```
[@mrdoob]
```
or
```
[                   @mrdoob                                ]
```
?

With the help of this PR, I will be able to call you by EITHER name depending on your preference.

**Comment by donmccurdy** [comment]
Following up on discussion in https://github.com/mrdoob/three.js/pull/17940, a few comments:

The behavior under `minimal = false` here considers morph targets; the behavior under `minimal = true` does not. I don't think that difference is captured in the name "minimal" ... should it consider morph targets? If so, perhaps it should consider the _current state_ of the morph targets, rather than the _maximum possible extent_ of the morph targets (as `computeBoundingBox` does)?

@mrdoob had made another suggestion in https://github.com/mrdoob/three.js/pull/17940#issuecomment-611938083, for creating a new method `expandByObjectBounds`. Under that naming, presumably the behavior would be:

- **expandByObjectBounds / setFromObjectBounds:** Expands the Box3 conservatively and efficiently, based on a cached geometry bounding box that considers the _maximum possible_ extent of morph targets, if present.
- **expandByObject / setFromObject:** Expands the Box3 precisely and at greater expense, based on vertex data and the _current state_ of morph targets, if present.

In terms of clarity, I might prefer that API. The downside is that it decreases performance on the existing method, with little way to communicate that to users. Brainstorming a few alternatives...

```js
// names indicating use of `geometry.boundingBox` and maximum morph target extent...
expandByObject( object, useBounds = true )
expandByObject( object, boundingBox = true )
expandByObject( object, cached = true )
expandByObject( object, useMaximumExtent = true )

// names indicating a full scan of vertex data and current morph target state...
expandByObject( object, deep = false )
expandByObject( object, exact = false )
expandByObject( object, tight = false )
```

It's also possible that the geometry's bounding box was computed without considering morph targets, or at a much larger or smaller size (e.g. to account for skinning bind matrices). So I think it's probably good to be explicit about use of the bounding box rather than to use a word like "tight" or "minimal".

**Comment by mrdoob** [comment]
> ```js
> expandByObject( object, exact = false )
> ```

That one sounds good to me.

As well as this one:

```js
expandByObject( object, precise = false )
```

**Comment by mrdoob** [comment]
@arikwex Sorry for the huge delay... 😬

Do you mind resolving the merge conflicts?

**Comment by arikwex** [comment]
Bones rattle in a deep and dark cave. A slumbering beast awakens, hungry for a bounding box. It rebases...

Everything should be up-to-date and the signature for the expandByObject + setFromObject methods have been adjusted with the name `precise` instead of `minimal`. However, I haven't changed the morphTargets. I couldn't tell from @donmccurdy's comment if that's a requirement to merge this change.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**